### ISSUE TITLE
Spec: we need to solve non-determinism for floating point columns

### ISSUE THREAD

**Issue by jsign** [tablelandnetwork/docs#27]
This is just a mirror issue for the discussion happening [in Notion](https://www.notion.so/textile/Floating-point-math-is-hard-7c3e39f4062d4698a29584bb24655b8e).

**Comment by carsonfarmer** [comment]
Just going to pull some of the linked Notion discussion here for visibility.

**Comment by carsonfarmer** [comment]
**TL;DR** Floating point math is, by definition an approximation, so how are we going to deal with that in a blockchain scenario where a network of validators abhors non-determinism!

# The problem

Recently, our core detectives (ahem, developers) discovered an issue with floating point math when testing our validator network: [How we detected a surprising non-deterministic cause of execution](https://www.notion.so/How-we-detected-a-surprising-non-deterministic-cause-of-execution-97697d202bf54be6a9c04f854b10f91b). In the subsequent research, they even found a reference to an article that [doubts we can trust floating point math on Apple silicon](https://eclecticlight.co/2021/04/22/can-you-trust-floating-point-arithmetic-on-apple-silicon/). That’s not great, but it isn’t entirely surprising, because it turns out [floating point math is hard](https://docs.oracle.com/cd/E19957-01/806-3568/ncg_goldberg.html). Lots of [people know this](https://riskledger.com/blog/floating-point-numbers/), which is probably why they aren’t used for storing currencies or with blockchains very often.

# Some “solutions”

- Don’t support floating point values (and suggest something else like text)
- ~~Forcefully truncate things to some reasonable precision that we *can* safely support~~
- ~~Just live with it, and expect validators to only support specific architectures/hardwares~~
- Find some code-based solution to work around the issue
- ~~Others?~~

# Some initial thoughts

I’m really *not* a fan of *not* supporting `FLOAT`. This is a pretty big DX step backward. Additionally, expecting people to think about integer math is asking a lot. Which reminds me of another issue. This discussion is actually a lot bigger than just *representing* floating point numbers… it is also about **floating point math**. If folks have two numbers in a table, they are *going* to want to do math on those numbers, and *probably* in the database. So something like…

```sql
select cast(5 as int) / cast(2 as int);
-- 2
select cast(5 as float) / cast(2 as float);
-- 2.5
```

… seems like a pretty intuitive expectation for most devs, and something they should be able to do. Note that the even *more* intuitive behavior of…

```sql
create table math_is_hard (a_r real, a_i int, b_r real, b_i int) strict; -- added strict here just to make sure i'm doing it right
insert into math_is_hard values (5, 5, 2, 2);
select a_r / b_r from math_is_hard;
-- 2.5
select a_i / b_r from math_is_hard;
-- 2.5
select a_i / b_i from math_is_hard;
-- 2
select a_r / b_i from math_is_hard;
-- 2.5
```

… is slightly less explicit, but equally valid if they’re into analytics and not computer science!

<aside>
🤔 While we’re at it, representing big integers is also not supporting in SQLite, and this is starting to look like an issue as well. Right now we’re recommending folks adopt an alternatively representation (text). We *can* help to address this using our SDK (opaquely convert BigInt to text on ingest), and we *can* do something similar with floats (e.g., `Number.isInteger` with Javascript helps in the SDK).

</aside>

## What does SQLite have to say about this?

With all that in mind, here’s what SQLite docs have to say about floating point numbers:

> SQLite stores integer values in the 64-bit [twos-complement](https://en.wikipedia.org/wiki/Two%27s_complement) format. This gives a storage range of -9223372036854775808 to +9223372036854775807, inclusive. Integers within this range are exact.

So-called "REAL" or floating point values are stored in the [IEEE 754](https://en.wikipedia.org/wiki/IEEE_754) [Binary-64](https://en.wikipedia.org/wiki/Double-precision_floating-point_format) format. This gives a range of positive values between approximately 1.7976931348623157e+308 and 4.9406564584124654e-324 with an equivalent range of negative values. A binary64 can also be 0.0 (and -0.0), positive and negative infinity and "NaN" or "Not-a-Number". Floating point values are approximate.

Pay close attention to the last sentence in the previous paragraph:

    **Floating point values are approximate.**

If you need an exact answer, you should not use binary64 floating-point values, in SQLite or in any other product. This is not an SQLite limitation. It is a mathematical limitation inherent in the design of floating-point numbers.

Furthermore, “SQLite promises to preserve the 15 most significant digits of a floating point value. However, it makes no guarantees about the accuracy of computations on floating point values, as no such guarantees are possible.” This might suggest that truncation is a totally reasonable thing for us to do?

The docs do mention that there are [extensions for dealing with floating point numbers](https://www.sqlite.org/floatingpoint.html#extensions_for_dealing_with_floating_point_numbers), but again, these are based on alternative representations of floating point numbers, namely as text. They are a good idea, and we *could* support these in our Validators (they aren’t included by default), and add docs to suggest that folks leverage this.

But at the end of the day, we do still need to deal with the fact that **floating point math is hard**.

**Comment by carsonfarmer** [comment]
I'm not going to mirror the discussions from that Notion doc (I can provide a summary if that is helpful later). However, the outcome of that was essentially that we _can't really support REAL values in our spec safely_.

**Comment by carsonfarmer** [comment]
Thanks to some initial sleuthing by @jsign we know that there are only really 87 "tables" that leverage REAL column data types, and after deduplication, that leaves us with really only ~41 tables affected. Some of these tables appear to be stale, or unused, and some should be pretty easy to identify their owners.

**Comment by carsonfarmer** [comment]
For those following along, the proposal is now to **drop support for REAL data types in our spec**. If you have an issue with this proposal/decision, please speak now or forever hold your peace.

**Comment by carsonfarmer** [comment]
Ok, so I'm going to consider this discussion "passed" and I'm going to address it with a PR to _this repo_ with the changes to the spec to reflect the fact that we no longer support REAL values in our spec.

**Comment by brunocalza** [comment]
@carsonfarmer one thing we did not explicitly discuss it is that we have to drop also support for literal float number. For example, `SELECT 2.333 / column FROM table`. The `2.333` is a literal float as is subject to the same problem.

**Comment by asutula** [comment]
Just curious why any of these rules to avoid non-determinism matter for read statements? Haven't been a part of the conversation, so I'm probably just not aware of something basic. Just read that last message and it's counter intuitive to me. I'd imagine that determinism is only required for writing data to tables and for read queries that are used to validate state (which the protocol should define and have control over).

**Comment by brunocalza** [comment]
Good point. We haven't really discussed read queries tbh. Most of the discussion revolved around the storage class types. But my take on that is that one of the things we want to guarantee is that different gateways return the exact same data. Even though the data stored can be exactly the same, we cannot guarantee with float operations on that data that the result will be the same across different validators.

We face the same problem whether SQLite is doing float operation before write or after write.

**Comment by brunocalza** [comment]
Just throwing another kind of unrelated comment. I think this is related to why we need two kinds of proof. One is proof of the correct stored state and another is proof of retrieval/queries.

**Comment by carsonfarmer** [comment]
Also another reason is perhaps just avoiding confusion. If we allow things like FLOAT literals in read queries, someone is going to ask "why here but not there", or "why do I get this result from one validator, and this from another". Better to just avoid REAL values altogether and add them back later if we find a solution.

**Comment by asutula** [comment]
Ok good points. I hadn't considered the importance of consistent query results from different validators or proof of retrieval.
